drivers/generic/generic.go

```diff
@@ -0,0 +1,269 @@
+	"github.com/codegangsta/cli"
+	"github.com/docker/machine/drivers"
+	"github.com/docker/machine/ssh"
+	"github.com/docker/machine/state"
+	"github.com/docker/machine/utils"
+)
+
+const (
+	dockerConfigDir = "/etc/docker"
+)
+
+type Driver struct {
+	MachineName    string
+	IPAddress      string
+	User           string
+	SSHKey         string
+	SSHPort        int
+	CaCertPath     string
+	PrivateKeyPath string
+	DriverKeyPath  string
+	storePath      string
+}
+
+func init() {
+	drivers.Register("generic", &drivers.RegisteredDriver{
+		New:            NewDriver,
+		GetCreateFlags: GetCreateFlags,
+	})
+}
+
+// GetCreateFlags registers the flags this driver adds to
+// "docker hosts create"
+func GetCreateFlags() []cli.Flag {
+	return []cli.Flag{
+		cli.StringFlag{
+			Name:  "generic-ip-address",
+			Usage: "IP Address of machine",
+			Value: "",
+		},
+		cli.StringFlag{
+			Name:  "generic-user",
+			Usage: "User to use for machine",
+			Value: "root",
+		},
+		cli.StringFlag{
+			Name:  "generic-ssh-key",
+			Usage: "SSH private key path",
+			Value: "",
+		},
+		cli.IntFlag{
+			Name:  "generic-ssh-port",
+			Usage: "SSH port",
+			Value: 22,
+		},
+	}
+}
+
+func NewDriver(machineName string, storePath string, caCert string, privateKey string) (drivers.Driver, error) {
+	return &Driver{MachineName: machineName, storePath: storePath, CaCertPath: caCert, PrivateKeyPath: privateKey}, nil
+}
+
+func (d *Driver) DriverName() string {
+	return "generic"
+}
+
+func (d *Driver) SetConfigFromFlags(flags drivers.DriverOptions) error {
+	d.IPAddress = flags.String("generic-ip-address")
+	d.User = flags.String("generic-user")
+	d.SSHKey = flags.String("generic-ssh-key")
+	d.SSHPort = flags.Int("generic-ssh-port")
+
+	if d.IPAddress == "" {
+		return fmt.Errorf("generic driver requires the --generic-ip-address option")
+	}
+
+	if d.SSHKey == "" {
+		return fmt.Errorf("generic driver requires the --generic-ssh-key option")
+	}
+
+	return nil
+}
+
+func (d *Driver) PreCreateCheck() error {
+	return nil
+}
+
+func (d *Driver) Create() error {
+	log.Infof("Importing SSH key...")
+
+	if err := utils.CopyFile(d.SSHKey, d.sshKeyPath()); err != nil {
+		return fmt.Errorf("unable to copy ssh key: %s", err)
+	}
+
+	if err := os.Chmod(d.sshKeyPath(), 0600); err != nil {
+		return err
+	}
+
+	log.Infof("Waiting for SSH...")
+
+	if err := ssh.WaitForTCP(fmt.Sprintf("%s:%d", d.IPAddress, d.SSHPort)); err != nil {
+		return err
+	}
+
+	log.Info("Configuring machine...")
+
+	log.Debugf("Setting hostname: %s", d.MachineName)
+	cmd, err := d.GetSSHCommand(fmt.Sprintf(
+		"echo \"127.0.0.1 %s\" | sudo tee -a /etc/hosts && sudo hostname %s && echo \"%s\" | sudo tee /etc/hostname",
+		d.MachineName,
+		d.MachineName,
+		d.MachineName,
+	))
+
+	if err != nil {
+		return err
+	}
+	if err := cmd.Run(); err != nil {
+		return err
+	}
+
+	log.Debug("Installing Docker")
+
+	cmd, err = d.GetSSHCommand("if [ ! -e /usr/bin/docker ]; then curl get.docker.io | sudo sh -; fi")
```

Comment:
this seems like a number of things could go wrong and it would be hard to know what it was... the curl could fail or you might not have sudo permissions...